### ISSUE TITLE
program panics when default_tgs_enctypes has unsupported enctype

### ISSUE THREAD

**Issue by dpotapov** [jcmturner/gokrb5#50]
Hi,

If I have unsupported enctype in the first place of my krb5.conf, the program panics:

```
default_tgs_enctypes = des-cbc-crc rc4-hmac
```

It panics because there is no error check in `krb5Token.go` line 130 and `etype` is `nil`:

```
	etype, _ := crypto.GetEtype(keyType)
	auth.GenerateSeqNumberAndSubKey(keyType, etype.GetKeyByteSize())
```

Is it possible to ignore unsupported enctypes in `default_tgs_enctypes` config parameter?

Thanks in advance!

**Comment by jcmturner** [comment]
Hi, can you share some more details on your set up and code. I tried to recreate the issue but was unable to get it to panic as you were seeing. Nevertheless not checking the error is bad of me so I've rectified that anyway.

**Comment by dpotapov** [comment]
ok, so I made some investigation with Wireshark, hope the info below will help (maybe for #58 as well).

My initial code is pretty straightforward:

```
cfg, _:= config.Load("C:\\temp\\krb5.conf")
cl := client.NewClientWithPassword("user", "AAA.COM", "P@$$w0rd")
cl.GoKrb5Conf.DisablePAFXFast = true
cl.WithConfig(cfg)

cl.Login()       // Login works
...
cl.SetSPNEGOHeader(r, "")     // throws error "Error getting etype to encrypt authenticator: [Unknown or unsupported EType: 1]"
```

When `krb5.conf` have lines like I mentioned above:
```
default_tgs_enctypes = des-cbc-crc rc4-hmac
default_tkt_enctypes = des-cbc-crc rc4-hmac
```
the client sends an "AS-REQ" packet to KDC with 2 enc types: 23 and 1 (in that order!).

KDC replies with KRB Error message (PREAUTH_REQUIRED) that includes PA-DATA structure with following entries (which seems to be a correct behaviour):
* PA-ENCTYPE-INFO, which has only one ETYPE-INFO-ENTRY with etype 23 and no salt.
* PA-ENCTYPE-INFO2, it also has only one ETYPE-INFO2-ENTRY with etype 23
* PA-ENC-TIMESTAMP
* PA-DASS
* PA-PK-AS-REP

Then the client sends "TGS-REQ" packet with 2 enc types in following order: 1 and 23.
KDC responds with "TGS-REP" packet that contains authenticator encrypted with etype = 1 (des-cbc-crc), though other fields are encrypted with etype = 23 (rc4-hmac).

Basically ordering of etypes in AS-REQ/TGS-REQ matters.

I've found following lines in `ASExchange.go`:
```
sort.Sort(sort.Reverse(sort.IntSlice(cl.Config.LibDefaults.DefaultTktEnctypeIDs)))
etype, err := crypto.GetEtype(cl.Config.LibDefaults.DefaultTktEnctypeIDs[0])
```

Just curious, is this sorting method described in any RFC?

Anyway applying that code in `KDCReq.go` for `NewTGSReq()` resolves the issue.

**Comment by bodgit** [comment]
The `krb5.conf(5)` man page says the `default_*_enctypes` values are listed in order of preference from highest to lowest.

So to my eyes, the order should be respected rather than sort them again by ID (in descending order, if I'm reading the code correctly) as that would naturally put RC4-HMAC (and Camellia, currently unsupported) at the front of the list as they have the highest ID, and I think I'm correct in saying use of RC4-HMAC at least should be discouraged.

**Comment by jcmturner** [comment]
@dpotapov Thanks for the detailed analysis. As @bodgit points out the ordering is important and something I'll take a look at. I need to think about why I did this in the first place.
